### PR TITLE
Destroy process group in DDP destructor

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Discovered by running a DeepSpeed test without `special=True` which would produce the error:

`RuntimeError: a leaf Variable that requires grad is being used in an in-place operation.`

This is because a previous test initialized the process group with the `gloo` backend and we never destroyed it because it raised a `SystemExit` on setup. And since special tests are run separately this wasn't caught before.

https://github.com/PyTorchLightning/pytorch-lightning/blob/f79f0f9de112bc0a810587147b2ff5d1f3bfb256/tests/accelerators/test_ddp.py#L98-L122

So the fix is:

- Properly release the group in ~~teardown~~ (edit: destructor, we want to reuse the process group across `fit/validate/test` calls) 
- ~~Do not raise a SystemExit in setup as teardown won't get called in that case. We should eventually also clean up after ourselves in cases like that.~~

The changes in #8070 need this to work so that will act as a test

Closes #8115 

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

**Review by SeanNaren — APPROVED** [review]

**Review by Borda — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM ! Great catch !

**Comment by SeanNaren** [comment]
@awaelchli caught the fact that we probably want to keep the communication alive till the end of all trainer stages. Not entirely sure what would be the preferred approach if this is the case!

File: CHANGELOG.md [hunk 1]
```diff
@@ -285,6 +285,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Support manual optimization with DeepSpeed ([#7970](https://github.com/PyTorchLightning/pytorch-lightning/pull/7970))
 
 
+- Destroy the distributed process group on DDP teardown ([#8080](https://github.com/PyTorchLightning/pytorch-lightning/pull/8080))
```
Comment (carmocca) [inline CHANGELOG.md:288]:
```suggestion
- Destroy the distributed process group on DDP destructor ([#8080](https://github.com/PyTorchLightning/pytorch-lightning/pull/8080))
```

File: pytorch_lightning/plugins/training_type/ddp.py [hunk 2]
```diff
@@ -373,3 +373,7 @@ def register_plugins(cls, plugin_registry: Dict) -> None:
             description="DDP Plugin with `find_unused_parameters` as False",
             find_unused_parameters=False
         )
+
+    def __del__(self) -> None:
+        if torch_distrib.is_initialized():
+            torch_distrib.destroy_process_group()
```
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/ddp.py:379]: @awaelchli do you think we should still call `self.teardown()` here?

which is the parent's:

```python
        if self.on_gpu:
            # GPU teardown
            self.lightning_module.cpu()
            # clean up memory
            with torch.cuda.device(self.root_device):
                torch.cuda.empty_cache()
```

maybe just the empty cache?

Comment (awaelchli) [reply pytorch_lightning/plugins/training_type/ddp.py:379]: I don't know if process group leaves memory in cache. probably it does. 
but I think teardown generally should only run once per fit. so we might just want to call empty_cache() in del.

**Comment by awaelchli** [comment]
@ananthsub is destroying the process group on trainer destruction breaking any use cases you know of?

for example, this will create and destroy the group N times, before it would only create it once:

```python
for n in range(N):
   trainer = Trainer(max_epochs=n, ...):
   trainer.fit(...)
   ...
```

File: pytorch_lightning/plugins/training_type/ddp.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/plugins/training_type/ddp.py:379]:
```suggestion
    def __del__(self) -> None:
        if torch_distrib.is_initialized():
            torch_distrib.destroy_process_group()
            # clean up memory
            with torch.cuda.device(self.root_device):
                torch.cuda.empty_cache()
```

**Comment by carmocca** [comment]
> this will create and destroy the group N times

So this fix is necessary because if the trainer config is different each iteration (maybe a different plugin is used), side effects from the previous process group can impact the new trainer.

File: pytorch_lightning/plugins/training_type/ddp.py (hunk 2 above)
Comment (justusschock) [reply pytorch_lightning/plugins/training_type/ddp.py:379]: I think we should do this (although I think when the process group is destroyed, it's GPU memory should also be freed since there is nothing holding it anymore)

**Review by justusschock — APPROVED** [review]

**Comment by ananthsub** [comment]
@tchaton - what if `trainer.save_checkpoint` is used outside of the the top-level trainer functions (fit/test/etc) ? Wouldn't this break if saving the checkpoint required the process group for syncing?

File: pytorch_lightning/plugins/training_type/ddp.py (hunk 2 above)
Comment (ananthsub) [reply pytorch_lightning/plugins/training_type/ddp.py:386]: n00b question: why do the process group destruction here vs in the training type plugin's teardown?

**Comment by carmocca** [comment]
> what if trainer.save_checkpoint is used outside of the the top-level trainer functions (fit/test/etc) ? Wouldn't this break if saving the checkpoint required the process group for syncing?

As long as you have the trainer reference alive, `__del__` won't be called for the `DDPPlugin` as both the `Accelerator` and `AcceleratorConnector` classes hold strong references to the training type plugin.

So if you are able to call `trainer.save_checkpoint`, the process group won't be destroyed.

**Review by ananthsub — COMMENTED** [review]
@carmocca what if the user creates the process group before any call to the trainer? they'd expect it to be around after the trainer is destroyed too. if it's destroyed as part of the trainer destructor, this might break expectations. should the destruction only happen if the plugin itself created the process group?

more advanced use cases (like mixture of experts or other sharded models) will require exposing the process group to the lightning module in order to define the model architecture before training. in such cases, what guarantees can we make regarding the lifetime of that process group?

**Comment by justusschock** [comment]
@ananthsub I think that this is a good compromise: Everything that was created by the trainer/plugin can/will also be destroyed by the trainer/plugin and we do not give any guarantees on its lifetime beyond the trainer/plugin's lifetime.

When they are created by the trainer, they usually are just necessary for training/validation/testing/prediction and could be safely deleted afterwards. If the user wants them to persist beyond the trainer, they have to create them manually beforehand (We could also introduce a flag for that, but I would avoid that in general, since to me it seems like a good general pattern that we try to cleanup everything we created - This accounts for everything, not just processes  ut most things are cleaned automatically upon destruction).

**Comment by awaelchli** [comment]
not disagreeing, just a note that the whole process group logic depends on a global state in the torch library so what we clean up here is not necessarily what we created.
  
so the safest for the user is to probably create their own process group, one that is different from the default one, and we only handle the default? does that make sense?

**Comment by justusschock** [comment]
@awaelchli You're right, that is why we need to ensure that (e.g. by setting an internal flag when creating it). That also does not mean that we are killing the processes, we just don't provide the communication interface anymore. So the user is free to setup the communication again after the trainer finishes/is destroyed.

File: pytorch_lightning/plugins/training_type/ddp.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/plugins/training_type/ddp.py:386]: > why do the process group destruction in the plugin's destructor vs vs in the training type plugin's teardown?

Quoting Adrian:

> but we don’t want to destroy the group after fit. it needs to stay alive until the process ends.
> we want to run more after the first Trainer.fit(), like Trainer.test() validate() or predict()

> if someone instantiates a plugin object and passes it into the trainer, it's still live after the trainer.fit call, so the process group won't be destroyed right? then wouldn't subsequent calls to the trainer could still observe the issue mentioned in the PR summary?

This would only be an issue when the second trainer call uses a different distributed process group configuration to the first trainer call. This would be an extreme edge case for users. But it is not on our CI where pytests runs the tests run on the same process

**Comment by carmocca** [comment]
Status update on this PR. Spent a ton of time debugging several issues. Here is the recap of all the options explored and their pros and cons

## Option 1: destroy on `teardown`

Pros:
- Simple solution
- Well-defined behaviour

Cons:
- No process group when `fit` stops which would break BC as users currently rely on the default process group
- Process group is recreated each `fit` call

### Option 1b: Use a custom group and destroy on `teardown`

Pros:
- Allows users to create and use their own process groups. This avoids conflicts with our own.

Cons:
- Any collective ops will need to pass the group if they are not done through the model. e.g.

```python
 def training_step():
        ...
        torch.distributed.allgather(...)  # would need to pass the group after this change
        self.model.allgather()
```

## Option 2: destroy in `conftest.py` fixture

Pros:
- Simplest solution.
- The only one entirely backwards compatible

Cons:
- Ugly solution
- Doesnt address proper resource management

Explored in https://github.com/PyTorchLightning/pytorch-lightning/pull/8115/
 
## Option 3: destroy in `__del__`

Pros
- Most efficient solution.

Cons:
- No process group when the trainer is destroyed
- NCCL errors. Potentially need to run CI on torch 1.8. Explored in #8116 
- It's up to the GC to decide when to delete the classes. So we need to manually run it in a few tests:

```python
def fn():
    trainer = ...  # with `DDPPlugin`
    trainer.fit()

fn()
gc.collect()  # necessary otherwise `__del__` could be called after the init call and the code would break
torch.distributed.init_process_group()
```

This PR currently implements option (3)

**Comment by carmocca** [comment]
@justusschock (requested review again since) @awaelchli 

This should be finally ready

File: pytorch_lightning/plugins/training_type/ddp.py [hunk 3]
```diff
@@ -304,7 +304,7 @@ def init_ddp_connection(self, global_rank: Optional[int] = None, world_size: Opt
         world_size = world_size if world_size is not None else self.cluster_environment.world_size()
         os.environ["MASTER_ADDR"] = self.cluster_environment.master_address()
         os.environ["MASTER_PORT"] = str(self.cluster_environment.master_port())
-        if not torch.distributed.is_initialized():
+        if not torch_distrib.is_initialized():
```
Comment (awaelchli) [inline pytorch_lightning/plugins/training_type/ddp.py:307]: can we just keep the standard name? it seems so arbitrary to save 3 characters

File: tests/plugins/test_deepspeed_plugin.py [hunk 4]
```diff
@@ -265,6 +266,10 @@ def backward(self, loss: Tensor, optimizer: Optimizer, optimizer_idx: int, *args
                                                     (RandomIterableDataset, "auto"), (RandomIterableDataset, 10)])
 def test_deepspeed_auto_batch_size_config_select(tmpdir, dataset_cls, value):
```
Comment (awaelchli) [inline tests/plugins/test_deepspeed_plugin.py:267]: I bet this would fail if we set gpus > 1 due to the parameterization :) haha

**Review by awaelchli — COMMENTED** [review]
I am still convinced that option 2 is the best solution, but I'm also ok with the current one. 

The only unknown for me is if relying on garbage collection is safe when dealing with concurrency. After @carmocca explained to me why we needed to call `gc.collect()` in some of the tests, one question came to my mind.

What happens if the automatic GC is faster on one process than the other? Is this possible in theory? If yes then I think we can come up with an example where we hang or crash unexpectedly.

For example:

```python
def train()
   trainer = Trainer()
   trainer.fit()
   
def test()
    trainer = Trainer()
    trainer.test()
    
train()
test() 
```

Imagine this code is executed by two processes. 

One process could theoretically enter `trainer.test()` in foo_1() without GC having destroyed the old Trainer. But if the other process is faster with GC, it will destory the process group and then enter `trainer.test()` with a call to init a new process group. 
Now if my theory is correct there will be a hang, because all processes would need to enter `init_process_group()`.

File: pytorch_lightning/plugins/training_type/ddp.py (hunk 3 above)
Comment (carmocca) [reply pytorch_lightning/plugins/training_type/ddp.py:307]: The rest of the file uses this abbreviation so I was just being consistent.

Can update all the occurrences

File: tests/plugins/test_deepspeed_plugin.py (hunk 4 above)
Comment (carmocca) [reply tests/plugins/test_deepspeed_plugin.py:267]: That could explain why some work with `gc.collect` and others do not

Comment (awaelchli) [reply tests/plugins/test_deepspeed_plugin.py:267]: yep! let's check tomorrow if we see a pattern across the special tests :)

File: pytorch_lightning/plugins/training_type/ddp.py (hunk 3 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/training_type/ddp.py:307]: aha ok, then whatever you feel is better.

**Comment by carmocca** [comment]
> One process could theoretically enter trainer.test() in test() without GC having destroyed the old Trainer yet.

Yes, thinking just single-process, the example would need to run `gc.collect()` in between the `train()` and `test()` calls to make sure `__del__` does not run after the check to `init_process_group` in `trainer.test()`

> because all processes would need to enter init_process_group()

Is that the case?

Then why do we guard the `init_process_group` call under `is_initialized`? Wouldn't that mean if one process initializes it before the other runs the `if`  statement, then the other wouldn't reach `init_process_group` and hang?

File: pytorch_lightning/plugins/training_type/ddp.py (hunk 3 above)
Comment (carmocca) [reply pytorch_lightning/plugins/training_type/ddp.py]: Done in 7a123543ab23c70920f9528a413479690a5d898e

**Comment by awaelchli** [comment]
> Then why do we guard the init_process_group call under is_initialized?

because in this case here we don't want to destroy and recreate the process group (option 3 in this pr):

```python
trainer = Trainer()
trainer.fit() # not yet initialized
trainer.test() # already initialized
```

if we didn't have the guard we would essentially have to go for option 1, right?

File: pytorch_lightning/accelerators/gpu.py [hunk 5]
```diff
@@ -42,10 +42,7 @@ def setup(self, trainer: 'pl.Trainer', model: 'pl.LightningModule') -> None:
 
     def on_train_start(self) -> None:
         # clear cache before training
-        # use context because of:
-        # https://discuss.pytorch.org/t/out-of-memory-when-i-use-torch-cuda-empty-cache/57898
-        with torch.cuda.device(self.root_device):
-            torch.cuda.empty_cache()
+        torch.cuda.empty_cache()
```
Comment (justusschock) [inline pytorch_lightning/accelerators/gpu.py:45]: We did set the default device before already, right?

**Review by justusschock — APPROVED** [review]
LGTM overall. Just a general question:

There are several `torch.cuda.empty_cache` calls in parts that should be hardware agnostic (like different trainingtype plugins etc.) while I know it doesn't hurt since they have an internal guard, do you think we could/should move them somehow to the accelerator so that these parts really become agnostic?

File: pytorch_lightning/accelerators/gpu.py (hunk 5 above)
Comment (carmocca) [reply pytorch_lightning/accelerators/gpu.py:45]: https://github.com/PyTorchLightning/pytorch-lightning/blob/55a90af7fc0805855684e93dbad669f5bbe76eee/pytorch_lightning/accelerators/gpu.py#L40

**Comment by carmocca** [comment]
> while I know it doesn't hurt since they have an internal guard, do you think we could/should move them somehow to the accelerator so that these parts really become agnostic?

If possible, yes

**Review by ananthsub — COMMENTED** [review]
I prefer option 2 (addressing this in the test fixture) to start because of the BC concerns @awaelchli mentioned. While we could track whether the plugin was the actor that inited the process group, we still wouldn't know if users call collective functions after the trainer is destroyed. @yifuwang what do you think? 

Meta question: For potential breaking changes like this, do we have an avenue to get feedback from other power users in the community?

**Comment by yifuwang** [comment]
@carmocca I think the plugin should only destroy the global pg if `init_process_group` is performed by the plugin. Otherwise, it doesn't own the global pg and should not destroy it.

**Comment by awaelchli** [comment]
@carmocca I think this is not needed anymore, right?

**Comment by carmocca** [comment]
@awaelchli We still don't destroy the process group after the `Trainer` ends. We do it in our CI through conftest.

However, it's unclear whether this is desired and even if we want to, it's a breaking change.

So I'm fine with closing this. Can always be revisited later.
